reimplementation.py

```diff
@@ -17,168 +19,280 @@
 
     """
-    df = read_df(filename, binarize)
-    # restructure data
-    classes_labels = np.array(df.index)
+    df, rv = read_df(filename, binarize)
+    rv_max = rv['replicate_value'].max()
+    class_labels = np.array(df.index)
     # penile skin should be treated separately
-    classes_set = set(classes_labels)
+    classes_set = set(class_labels)
     classes_set.remove('Skin.penile')
+
+    # initialize
     classes_map = {}
     inv_classes_map = {}
     i = 0
+
     for clas in sorted(classes_set) + ['Skin.penile']:
         if include_blank or 'Blank' not in clas:
             if not developing or ('Skin' not in clas and 'Blank' not in clas):
                 classes_map[clas] = i
                 inv_classes_map[i] = clas
                 i += 1
-    classes = [classes_map[clas] for clas in classes_labels if
+    classes = [classes_map[clas] for clas in class_labels if
                (not developing or ('Skin' not in clas and 'Blank' not in clas))
                and (include_blank or 'Blank' not in clas)]
-    n_per_class = Counter(classes)
+    n_per_class = Counter(classes) # This is not the final number per class!
+                                   # Samples can be discarded later on.
     n_features = len(df.columns)
-    X_raw = np.zeros([0, 4, n_features])
+
+    # Set the second dimension to rv_max so know for sure all measurements per
+    # sample will fit and get no dimension errors.
+    X_raw = np.zeros([0, rv_max, n_features])
     y = []
     for clas in sorted(classes_set) + ['Skin.penile']:
         if include_blank or 'Blank' not in clas:
             if not developing or ('Skin' not in clas and 'Blank' not in clas):
-                full_set_per_class = np.array(df.loc[clas])
-                # TODO find which make pairs of 4
-                # assuming they are ordered 4 together
-                n_samples, n_features = full_set_per_class.shape
-                print(clas, n_samples / 4)
-                # for now, discard the remainder?
-                n_full_samples = int(n_samples / 4)
-                n_per_class[classes_map[clas]] = n_full_samples
-                data_for_class = np.zeros((n_full_samples, 4, n_features))
+                # Implement which make pairs of replicates
+                data_for_this_label = np.array(df.loc[clas])
+                rv_set_per_class = np.array(rv.loc[clas]).flatten()
+                # Save the index when a 'new sample' starts and save in
+                # end_replicate. This is when the next integer is lower or equal
+                # than the current integer. (e.g. 4 > 1 or 2 == 2).
+                end_replicate = [ i for i in range(1, len(rv_set_per_class)) if
+                                  rv_set_per_class[i-1] > rv_set_per_class[i] or
+                                  rv_set_per_class[i-1] == rv_set_per_class[i]]
+                n_full_samples = len(end_replicate)+1
+                #TODO variabel aantal samples toestaan? --> wordt nu opgelost
```

Comment:
niet mooier en robuuster om het hier op te lossen? waarom is het geen probleem dat samples alleen maar lege markers kunnen bevatten?